### PR TITLE
Set-DbaLogin -Allow pipeline from Get-DbaLogin

### PR DESCRIPTION
**Pull request by awickham10**
## Type of Change

 - [ ] Bug fix (non-breaking change, fixes #<enter issue number>)
 - [x] New feature (non-breaking change, adds functionality)
 - [ ] Breaking change (effects multiple commands or functionality)
 - [x] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [x] Adding code coverage to existing functionality
 - [x] Pester test is included
 - [ ] Nunit test is included
 - [ ] Documentation
 - [ ] Build system
 

### Purpose
 
Accept Get-DbaLogin as input to Set-DbaLogin on the pipeline.

### Approach

Get-DbaLogin outputs SqlInstance, Name. On Set-DbaLogin I added ValueFromPipelineByPropertyName to the SqlInstance and Login parameters. I also had to add an alias of Name to the Login parameter.

### Commands to test

```powershell
Get-DbaLogin -SqlInstance sql1 -Login test | Set-DbaLogin -RemoveRole bulkadmin
```

### CODE REVIEW

**Comment by awickham10** [comment]
Dangit... had some branch issues, apparently I didn't get it fixed correctly. I'll give it another shot shortly.

**Comment by awickham10** [comment]
I think I got it switched over to the appropriate method (InputObject).

File: functions/Set-DbaLogin.ps1 [hunk 1]
```diff
@@ -130,14 +130,16 @@ function Set-DbaLogin {
...
-        [parameter(Mandatory, ValueFromPipeline)]
+        [parameter(Mandatory, ValueFromPipeline, ParameterSetName = "Instance")]
         [Alias("ServerInstance", "SqlServer")]
         [DbaInstanceParameter[]]$SqlInstance,
         [PSCredential]$SqlCredential,
-        [parameter(Mandatory)]
+        [parameter(Mandatory, ParameterSetName = "Instance")]
         [string[]]$Login,
+        [Parameter(ValueFromPipeline, Mandatory, ParameterSetName = "Logins")]
+        [Microsoft.SqlServer.Management.Smo.Login[]]$InputObject,
```
[the remark this answers is not in the document]
Comment (awickham10) [reply functions/Set-DbaLogin.ps1]: Got it

File: functions/Set-DbaLogin.ps1 [hunk 2]
```diff
@@ -130,14 +130,16 @@ function Set-DbaLogin {
 
 #>
 
-    [CmdletBinding()]
+    [CmdletBinding(DefaultParameterSetName = "Default")]
```
[the remark this answers is not in the document]
Comment (awickham10) [reply functions/Set-DbaLogin.ps1]: I see your point (parameter set errors are kind of meh), but having them makes the help documentation more clear and requires a bunch less manual validation that can probably be error prone. I had to add about 30 lines of manual validation to handle it and I'm still not 100% confident in the result.

I can submit the changes, but it seems a little less than optimal. Let me know what you think.

Code to Validate
image

Help Syntax Before with Parameter Sets - Note that something doesn't seem 100% right with this, so I still need to look at it, but it's nice that you can see your options right up front.
image

Help Syntax After
image

[the remark this answers is not in the document]
Comment (potatoqualitee) [reply functions/Set-DbaLogin.ps1]: the help is nice if people are looking at the syntax, but if our users are anything like me, theyll try to run a command, then get `Parameter set cannot be resolved using the specified named parameters`

as for validation, i'd be like `if ($SqlInstance -and $InputObject) { stop-function }`  or  `if ($SqlInstance -and -not $Login) { stop-function }`  or perhaps using our `Test-Bound` itnernal command.

but mostly, i dont see the harm in piping InputObjects plus specifying a sql instance to pile more on using `$InputObject +=`

[the remark this answers is not in the document]
Comment (wsmelton) [reply functions/Set-DbaLogin.ps1]: I see no purpose in having parameter sets for this command.

**Comment by wsmelton** [comment]
Test needs to also be updated to include the additional parameter.

**Comment by potatoqualitee** [comment]
added the requested updates and will merge when tests pass 💯

**Comment by potatoqualitee** [comment]
got it! todo: change to SMO Login output and update tests

**Comment by awickham10** [comment]
Thanks!!

**Pull request merged**